src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -0,0 +1,165 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "common.h"
+#include "frozenobjectheap.h"
+
+FrozenObjectHeapManager::FrozenObjectHeapManager():
+    m_CurrentHeap(nullptr)
+{
+    m_Crst.Init(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC);
+    m_HeapCommitChunkSize = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_FrozenSegmentCommitSize);
+    m_HeapSize = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_FrozenSegmentReserveSize);
+}
+
+// Allocates an object of the give size (including header) on a frozen segment.
+// May return nullptr in the following cases:
+//   1) DOTNET_FrozenSegmentReserveSize is 0 (disabled)
+//   2) Object is too large (large than DOTNET_FrozenSegmentCommitSize)
+// in such cases caller is responsible to find a more appropriate heap to allocate it
+Object* FrozenObjectHeapManager::AllocateObject(PTR_MethodTable type, size_t objectSize)
```

Comment:
Nit: Maybe this should be called `TryAllocateObject` to reflect the fact that it can return null.

Reply:
Same for `FrozenObjectHeap::AllocateObject`